### ISSUE TITLE
ERROR Could not execute template

### ISSUE THREAD

**Issue by JACONDE** [dynatrace-oss/dynatrace-monitoring-as-code#55]
I'm trying to execute a project called "General" that doesn't have any link with any other projects (./monaco -d -v -p="General" -e="./projects/tenants.yaml" projects) and when doing the "dry run", it seems to be doing a validation of another project called "Application", getting the following error message:

_```
" 2020-11-19 18:03:10 INFO  Dynatrace Monitoring as Code v1.0.0
2020-11-19 18:03:10 ERROR Could not execute template: template: projects\Application\app-detection-rule\rules.yaml:8:21: executing "projects\\Application\\app-detection-rule\\rules.yaml" at <.Env.URLPATTERN>: map has no entry for key "URLPATTERN"
2020-11-19 18:03:10 ERROR Error while converting file projects\Application\app-detection-rule\rules.yaml: template: projects\Application\app-detection-rule\rules.yaml:8:21: executing "projects\\Application\\app-detection-rule\\rules.yaml" at <.Env.URLPATTERN>:
map has no entry for key "URLPATTERN"
```_

**Comment by Dynatrace-Adam-Gardner** [comment]
A few things I notice here.

I'm assuming your folder structure is:

```
projects
> General
> > {Configuration Type Folders Are Here} 
```

First I see that your `Application` folder is at the wrong level. It's underneath `projects` but should actually live one level down under `General`.

Second, the error is complaining that you haven't set an environment varialbe called `URLPATTERN`.

```
set URLPATTERN=***  // Windows
// OR
export URLPATTERN=*** // Linux / Mac
```

Finally, be aware of https://github.com/dynatrace-oss/dynatrace-monitoring-as-code/issues/58 when naming your configuration type folders. I believe you need to rename the `Application` folder to `application`.

**Comment by JACONDE** [comment]
The current structure I have is:

projects
> General
> > {Configuration Type Folders Are Here} 
> Application
>> app-detection-rule
> > {Configuration Type Folders Are Here} 

I'm trying to execute the "General" project, but it is getting also validated the Application project, when there isn't any reference/link to that one in any of the General project .yaml files

**Comment by didiladi** [comment]
@JACONDE Sadly, we currently don't allow mixing of project folders with configuration folders. There is an explicit check in the tool to prevent this:

```
https://github.com/dynatrace-oss/dynatrace-monitoring-as-code/blob/main/pkg/project/project_loader.go#L176
```

If you just want to execute the `General` project, consider a structure like that:

```
projects
> General
> > application
> > app-detection-rule
> Application
> > alerting-profile
> > notification
```

**Comment by JACONDE** [comment]
thanks for the clarification !!

**Comment by Dynatrace-Adam-Gardner** [comment]
@JACONDE can you post your `projects\Application\app-detection-rule\rules.yaml` file please? It looks like you have a variable set like this:

```
config:
  - my_config: "some.json"

my_config:
  - someProperty: "{{ .Env.URLPATTERN }}"
```

In which case, the tool is expecting to have `URLPATTERN` set as an environment variable (see my previous post on how to set it).

**Comment by JACONDE** [comment]
Here is the info requested:
Sandbox_APP-0000:
  - name: "projects/Application/application/Sandbox_APP-0000.name"
  - applicationId: "projects/Application/application/Sandbox_APP-0000.id"
  - pattern: "{{ .Env.URLPATTERN }}"

Yes you are right. There is a variable called Env.URLPATTERN, However on the General project there isn't any reference to projects\Application\app-detection-rule, so not sure why it goes there to perform any validation.

**Comment by Dynatrace-Adam-Gardner** [comment]
It's due to your folder structure. At present (due to the capitalisation) monaco thinks you have two projects: `General` and `Application`. Application being nested inside `General`.

`monaco` doesn't allow nested projects so to fix just move the `Application` folder up one level so that it sits alongside (not underneath) `General`.

**Comment by JACONDE** [comment]
My project structure is:
projects
> General
> > application
> > others
> Application
> > alerting-profile
> > others

Not sure why when calling General project, it should validate also Application. Is this a bug?

**Comment by Dynatrace-Adam-Gardner** [comment]
Just rename `Application` to application and set the environment variable and this error will go away. It's not a bug. You've misnamed the folder. Configuration type folders should all be lowercase.

**Comment by JACONDE** [comment]
I just added the variables and it works. However not sure why I do need to have all environment variables for the entire project folder, in a pipeline for only one project

**Comment by UnseenWizzard** [comment]
Sorry for chiming in late here, but maybe I can clear things up a bit.

> why I do need to have all environment variables for the entire project folder, in a pipeline for only one project

If I followed the thread correctly, this is an environment variable you had used in the Application project? 

In general the tool reads in _all_ configs - at which point the env var resolution currently happens - in order to be able to handle dependencies between projects. 
That is admittedly an imperfect solution for having several projects, or in-progress projects, because as you noticed they all are validated... 

As I take it your original problem for this is resolved though @JACONDE ? 
Could you close this and possible create a feature request for not doing basic validation on unrelated projects?

**Comment by JACONDE** [comment]
Sure, I can do this. 
In the meantime, I think we should improve the documentation to make clear "tool reads in all configs - at which point the env var resolution currently happens - in order to be able to handle dependencies between projects"
Does it makes sense?

**Comment by tatomir146** [comment]
As this will be solved in #71 closing this issue.
